jni/patches/nmslib/0003-Added-streaming-apis-for-vector-index-loading-in-Hnsw.patch

```diff
@@ -0,0 +1,221 @@
++
+         Hnsw(bool PrintProgress, const Space<dist_t> &space, const ObjectVector &data);
+         void CreateIndex(const AnyParams &IndexParams) override;
+ 
+@@ -500,6 +502,7 @@ namespace similarity {
+ 
+         void SaveOptimizedIndex(std::ostream& output);
+         void LoadOptimizedIndex(std::istream& input);
++        void LoadOptimizedIndex(NmslibIOReader& input);
+ 
+         void SaveRegularIndexBin(std::ostream& output);
+         void LoadRegularIndexBin(std::istream& input);
+diff --git a/similarity_search/include/utils.h b/similarity_search/include/utils.h
+index b521c26..a3931b7 100644
+--- a/similarity_search/include/utils.h
++++ b/similarity_search/include/utils.h
+@@ -299,12 +299,24 @@ inline void WriteField(ostream& out, const string& fieldName, const FieldType& f
+   }
+ }
+ 
++struct NmslibIOReader {
++  virtual ~NmslibIOReader() = default;
++
++  virtual void read(char* bytes, size_t len) = 0;
++
++  virtual size_t remainingBytes() = 0;
++};
+ 
+ template <typename T> 
+ void writeBinaryPOD(ostream& out, const T& podRef) {
+   out.write((char*)&podRef, sizeof(T));
+ }
+ 
++template <typename T> 
++static void readBinaryPOD(NmslibIOReader& in, T& podRef) {
++  in.read((char*)&podRef, sizeof(T));
++}
++
+ template <typename T> 
+ static void readBinaryPOD(istream& in, T& podRef) {
+   in.read((char*)&podRef, sizeof(T));
+diff --git a/similarity_search/src/method/hnsw.cc b/similarity_search/src/method/hnsw.cc
+index 4080b3b..662f06c 100644
+--- a/similarity_search/src/method/hnsw.cc
++++ b/similarity_search/src/method/hnsw.cc
+@@ -950,7 +950,6 @@ namespace similarity {
+                   " read so far doesn't match the number of read lines: " + ConvertToString(lineNum));
+     }
+ 
+-
+     template <typename dist_t>
+     void
+     Hnsw<dist_t>::LoadRegularIndexBin(std::istream& input) {
+@@ -1034,6 +1033,143 @@ namespace similarity {
+ 
+     }
+ 
++    constexpr bool _isLittleEndian() {
++        return (((uint32_t) 1) & 0xFFU) == 1;
++    }
++                        
++    SIZEMASS_TYPE _readIntBigEndian(uint8_t byte0, uint8_t byte1, uint8_t byte2, uint8_t byte3) noexcept {
++        return (static_cast<SIZEMASS_TYPE>(byte0) << 24) |
++               (static_cast<SIZEMASS_TYPE>(byte1) << 16) |
++               (static_cast<SIZEMASS_TYPE>(byte2) << 8)  |
++               static_cast<SIZEMASS_TYPE>(byte3);
++    }                   
++                            
++    SIZEMASS_TYPE _readIntLittleEndian(uint8_t byte0, uint8_t byte1, uint8_t byte2, uint8_t byte3) noexcept {
++        return (static_cast<SIZEMASS_TYPE>(byte3) << 24) |
++               (static_cast<SIZEMASS_TYPE>(byte2) << 16) |
++               (static_cast<SIZEMASS_TYPE>(byte1) << 8)  |
++               static_cast<SIZEMASS_TYPE>(byte0);
++    } 
++
++    template <typename dist_t>
++    void Hnsw<dist_t>::LoadIndexWithStream(NmslibIOReader& input) {
++        LOG(LIB_INFO) << "Loading index from an input stream(NmslibIOReader).";
```

Comment:
[Question]:
is this info log enabled by default? Since I am not to see all the code from nmslib in this PR.

Reply:
It's disabled by default. (LIB_LOGNONE)

In nmslib_wrapper

```
void knn_jni::nmslib_wrapper::InitLibrary() {
  similarity::initLibrary();
}
```

and in NMSLIB
```
void initLibrary(int seed = 0, LogChoice choice = LIB_LOGNONE, const char*pLogFile = NULL);
```